File: android/app/src/main/kotlin/com/bockmedia/console/domain/model/SearchBrowse.kt
package com.bockmedia.console.domain.model

import com.bockmedia.console.data.api.dto.DashboardQuickResponse
import com.bockmedia.console.data.api.dto.GenreItem
import com.bockmedia.console.data.api.dto.AlbumItem
import com.bockmedia.console.data.api.dto.SearchHit
import com.bockmedia.console.data.repository.BockMediaRepository
import kotlinx.coroutines.async
import kotlinx.coroutines.coroutineScope
import kotlinx.coroutines.withTimeoutOrNull

data class SearchBrowseFeed(
    val pickedForYou: List<HomeCard> = emptyList(),
    val newReleases: List<AlbumItem> = emptyList(),
    val genres: List<GenreItem> = emptyList(),
)

enum class SearchSuggestionKind { Artist, Album, Playlist, Song }

data class SearchSuggestion(
    val kind: SearchSuggestionKind,
    val title: String,
    val subtitle: String? = null,
    val id: String? = null,
    val path: String? = null,
    val artist: String? = null,
    val album: String? = null,
)

object SearchBrowseLoader {
    private const val HISTORY_LIMIT = 40
    private const val PLAYLIST_LIMIT = 80
    private const val REQUEST_TIMEOUT_MS = 6_000L
    private const val ANALYTICS_TIMEOUT_MS = 3_000L

    /** Lightweight feed for first paint (genres + new releases). */
    suspend fun loadFast(repository: BockMediaRepository): SearchBrowseFeed = coroutineScope {
        val genresDef = async {
            withTimeoutOrNull(REQUEST_TIMEOUT_MS) {
                runCatching { repository.genres(limit = 16) }.getOrNull()
            }
        }
        val newReleasesDef = async {
            withTimeoutOrNull(REQUEST_TIMEOUT_MS) {
                runCatching { repository.recentAlbums(limit = 12) }.getOrNull()
            }
        }
        SearchBrowseFeed(
            genres = genresDef.await()?.items.orEmpty(),
            newReleases = newReleasesDef.await()?.items.orEmpty(),
        )
    }

    suspend fun load(repository: BockMediaRepository): SearchBrowseFeed = coroutineScope {
        val historyDef = async {
            withTimeoutOrNull(REQUEST_TIMEOUT_MS) {
                runCatching { repository.streamHistory(1, HISTORY_LIMIT) }.getOrNull()
            }
        }
        val analyticsDef = async {
            withTimeoutOrNull(ANALYTICS_TIMEOUT_MS) {
                runCatching { repository.analytics() }.getOrNull()
            }
        }
        val playlistsDef = async {
            withTimeoutOrNull(REQUEST_TIMEOUT_MS) {
                runCatching { repository.playlists(limit = PLAYLIST_LIMIT) }.getOrNull()
            }
        }
        val smartDef = async {
            withTimeoutOrNull(REQUEST_TIMEOUT_MS) {
                runCatching { repository.smartPlaylists() }.getOrNull()
            }
        }
        val genresDef = async {
            withTimeoutOrNull(REQUEST_TIMEOUT_MS) {
                runCatching { repository.genres(limit = 16) }.getOrNull()
            }
        }
        val newReleasesDef = async {
            withTimeoutOrNull(REQUEST_TIMEOUT_MS) {
                runCatching { repository.recentAlbums(limit = 12) }.getOrNull()
            }
        }
        val dashboardDef = async {
            withTimeoutOrNull(REQUEST_TIMEOUT_MS) {
                runCatching { repository.dashboardQuick() }.getOrNull()
            }
        }

        val history = historyDef.await()?.items.orEmpty()
        val analytics = analyticsDef.await()
        val allPlaylists = playlistsDef.await()?.items.orEmpty()
        val smartPlaylists = smartDef.await()?.items.orEmpty()
        val genres = genresDef.await()?.items.orEmpty()
        val newReleases = newReleasesDef.await()?.items.orEmpty()
        val dashboard = dashboardDef.await()

        val playlistByName = allPlaylists.associateBy { it.name.lowercase() }
        val playlistById = allPlaylists.associateBy { it.id }
        val usedArtPaths = mutableSetOf<String>()

        fun claimArt(path: String?): String? {
            if (path.isNullOrBlank() || path in usedArtPaths) return null
            usedArtPaths.add(path)
            return path
        }

        fun playlistCard(pl: com.bockmedia.console.data.api.dto.PlaylistSummary, artPath: String?, subtitle: String) =
            HomeCard(
                id = "pl-${pl.id}",
                title = pl.name,
                subtitle = subtitle,
                artPath = artPath,
                playlistId = pl.id,
                playTarget = PlayTarget.Playlist(pl.id, pl.name),
                kind = HomeSectionKind.Discover,
            )

        val recentPlaylistNames = linkedSetOf<String>()
        val artByPlaylist = mutableMapOf<String, String>()
        for (row in history) {
            row.playlist?.trim()?.takeIf { it.isNotBlank() }?.let { name ->
                recentPlaylistNames.add(name)
                row.filepath?.let { artByPlaylist.putIfAbsent(name.lowercase(), it) }
            }
        }

        val topGenres = analytics?.topGenres.orEmpty().take(6)
        val topArtists = analytics?.topArtists.orEmpty()

        val genreMixes = topGenres.mapIndexedNotNull { index, row ->
            val genre = row.name ?: row.label ?: return@mapIndexedNotNull null
            val smart = smartPlaylists.firstOrNull { sp ->
                !HomeFeedRules.isDailyMixName(sp.name) &&
                    (sp.name.contains(genre, ignoreCase = true) ||
                        sp.rules?.toString()?.contains(genre, ignoreCase = true) == true)
            }
            val linkedId = smart?.playlistId
            val artPath = claimArt(HomeFeedRules.artPathForGenreDistinct(history, genre, usedArtPaths))
                ?: claimArt(HomeFeedRules.artPathForArtistDistinct(
                    history,
                    HomeFeedRules.topArtistForGenre(history, genre) ?: topArtists.getOrNull(index)?.name.orEmpty(),
                    usedArtPaths,
                ))
                ?: claimArt(HomeFeedRules.nextDistinctArtPath(history, usedArtPaths))
            if (linkedId != null && playlistById[linkedId] != null) {
                playlistCard(playlistById[linkedId]!!, artPath, "$genre mix")
            } else {
                val named = allPlaylists.firstOrNull { HomeFeedRules.isGenreMixPlaylistName(it.name, genre) }
                if (named != null) {
                    playlistCard(named, artPath, "$genre mix")
                } else {
                    val seedArtist = HomeFeedRules.topArtistForGenre(history, genre)
                        ?: topArtists.getOrNull(index)?.name
                        ?: return@mapIndexedNotNull null
                    HomeCard(
                        id = "mix-$genre",
                        title = "$genre Mix",
                        subtitle = "Based on your listening",
                        artPath = artPath,
                        playTarget = PlayTarget.Artist(seedArtist),
                        kind = HomeSectionKind.TopMixes,
                    )
                }
            }
        }

        val recentAlbumCards = buildList {
            val seen = mutableSetOf<String>()
            for (row in history) {
                val album = row.album?.trim()?.takeIf { it.isNotBlank() } ?: continue
                val key = "${album.lowercase()}|${row.artist?.lowercase().orEmpty()}"
                if (!seen.add(key)) continue
                add(
                    HomeCard(
                        id = "album-$key",
                        title = album,
                        subtitle = row.artist ?: "Recently played",
                        artPath = claimArt(row.filepath),
                        playTarget = PlayTarget.Album(album, row.artist),
                        kind = HomeSectionKind.JumpBackIn,
                    ),
                )
                if (size >= 4) break
            }
        }

        val discoverCards = allPlaylists
            .filter { HomeFeedRules.isDiscoverName(it.name) }
            .sortedByDescending { parseSortDate(it.createDate) }
            .take(4)
            .map { pl ->
                playlistCard(
                    pl,
                    claimArt(artByPlaylist[pl.name.lowercase()]),
                    "Discover",
                )
            }

        val recentPlaylistCards = recentPlaylistNames.mapNotNull { name ->
            val pl = playlistByName[name.lowercase()] ?: return@mapNotNull null
            playlistCard(pl, claimArt(artByPlaylist[name.lowercase()]), "Played recently")
        }.take(4)

        val dashboardCards = dashboardRecentCards(dashboard, playlistByName, artByPlaylist, ::claimArt)
        val pickedForYou = (dashboardCards + genreMixes + recentAlbumCards + discoverCards + recentPlaylistCards)
            .distinctBy { it.id }
            .take(12)

        SearchBrowseFeed(
            pickedForYou = pickedForYou,
            newReleases = newReleases,
            genres = genres,
        )
    }

    suspend fun suggest(repository: BockMediaRepository, query: String): List<SearchSuggestion> {
        val q = query.trim()
        if (q.isEmpty()) return emptyList()
        if (q.length < 2) return suggestOneChar(repository, q)
        val response = runCatching { repository.search(q, limit = 8) }.getOrNull() ?: return emptyList()
        return suggestionsFromResponse(response)
    }

    suspend fun suggestOneChar(repository: BockMediaRepository, query: String): List<SearchSuggestion> =
        runCatching {
            repository.artists(1, query, limit = 6).items.map { artist ->
                SearchSuggestion(
                    kind = SearchSuggestionKind.Artist,
                    title = artist.name,
                    subtitle = "${artist.tracks} songs",
                )
            }
        }.getOrDefault(emptyList())

    fun suggestionsFromResponse(response: com.bockmedia.console.data.api.dto.SearchResponse): List<SearchSuggestion> =
        buildList {
            response.artists.take(2).forEach { hit ->
                add(
                    SearchSuggestion(
                        kind = SearchSuggestionKind.Artist,
                        title = hit.name ?: "",
                        subtitle = "Artist",
                        path = hit.path,
                    ),
                )
            }
            response.albums.take(2).forEach { hit ->
                add(
                    SearchSuggestion(
                        kind = SearchSuggestionKind.Album,
                        title = hit.name ?: "",
                        subtitle = hit.artist,
                        artist = hit.artist,
                        path = hit.path,
                    ),
                )
            }
            response.playlists.take(2).forEach { hit ->
                add(
                    SearchSuggestion(
                        kind = SearchSuggestionKind.Playlist,
                        title = hit.name ?: "",
                        subtitle = "Playlist",
                        id = hit.id,
                    ),
                )
            }
            response.songs.take(2).forEach { hit ->
                add(
                    SearchSuggestion(
                        kind = SearchSuggestionKind.Song,
                        title = hit.title ?: hit.name ?: "",
                        subtitle = hit.artist,
                        path = hit.path,
                        artist = hit.artist,
                        album = hit.album,
                    ),
                )
            }
        }.take(8)

    fun suggestionFromHit(kind: SearchSuggestionKind, hit: SearchHit): SearchSuggestion = when (kind) {
        SearchSuggestionKind.Artist -> SearchSuggestion(kind, hit.name ?: "", subtitle = hit.artist, path = hit.path)
        SearchSuggestionKind.Album -> SearchSuggestion(
            kind,
            hit.name ?: hit.album ?: "",
            subtitle = hit.artist,
            artist = hit.artist,
            path = hit.path,
        )
        SearchSuggestionKind.Playlist -> SearchSuggestion(kind, hit.name ?: "", id = hit.id)
        SearchSuggestionKind.Song -> SearchSuggestion(
            kind,
            hit.title ?: hit.name ?: "",
            subtitle = hit.artist,
            path = hit.path,
            artist = hit.artist,
            album = hit.album,
        )
    }

    fun playTargetFor(suggestion: SearchSuggestion): PlayTarget? = when (suggestion.kind) {
        SearchSuggestionKind.Artist -> PlayTarget.Artist(suggestion.title)
        SearchSuggestionKind.Album -> PlayTarget.Album(suggestion.title, suggestion.artist)
        SearchSuggestionKind.Playlist -> suggestion.id?.let { PlayTarget.Playlist(it, suggestion.title) }
        SearchSuggestionKind.Song -> suggestion.path?.let { PlayTarget.Song(it, suggestion.title) }
    }

    private fun parseSortDate(raw: String?): Long {
        if (raw.isNullOrBlank()) return 0L
        return runCatching { java.time.Instant.parse(raw).toEpochMilli() }.getOrElse {
            runCatching { java.time.ZonedDateTime.parse(raw).toInstant().toEpochMilli() }.getOrDefault(0L)
        }
    }
}

private fun dashboardRecentCards(
    dashboard: DashboardQuickResponse?,
    playlistByName: Map<String, com.bockmedia.console.data.api.dto.PlaylistSummary>,
    artByPlaylist: Map<String, String>,
    claimArt: (String?) -> String?,
): List<HomeCard> = dashboard?.recent.orEmpty().mapNotNull { item ->
    val playlistName = item.playlist?.trim()?.takeIf { it.isNotBlank() }
    if (playlistName != null) {
        val pl = playlistByName[playlistName.lowercase()] ?: return@mapNotNull null
        return@mapNotNull HomeCard(
            id = "dash-pl-${pl.id}",
            title = pl.name,
            subtitle = "Recently played",
            artPath = claimArt(artByPlaylist[playlistName.lowercase()] ?: item.path),
            playlistId = pl.id,
            playTarget = PlayTarget.Playlist(pl.id, pl.name),
            kind = HomeSectionKind.Discover,
        )
    }
    val path = item.path?.takeIf { it.isNotBlank() } ?: return@mapNotNull null
    val title = item.track?.takeIf { it.isNotBlank() } ?: return@mapNotNull null
    HomeCard(
        id = "dash-${path.hashCode()}",
        title = title,
        subtitle = item.artist ?: "Recently played",
        artPath = claimArt(path),
        playTarget = PlayTarget.Song(path, title),
        kind = HomeSectionKind.Discover,
    )
}

File: ios/BockMedia/Domain/SearchBrowse.swift
import Foundation

struct SearchBrowseFeed {
    var pickedForYou: [HomeCard] = []
    var newReleases: [AlbumItem] = []
    var genres: [GenreItem] = []
}

enum SearchSuggestionKind {
    case artist, album, playlist, song
}

struct SearchSuggestion: Identifiable {
    let kind: SearchSuggestionKind
    let title: String
    let subtitle: String?
    let hitId: String?
    let path: String?
    let artist: String?
    let album: String?

    var id: String {
        switch kind {
        case .artist: return "artist-\(title)"
        case .album: return "album-\(title)-\(artist ?? "")"
        case .playlist: return "pl-\(hitId ?? title)"
        case .song: return "song-\(path ?? title)"
        }
    }
}

enum SearchBrowseLoader {
    private static let historyLimit = 80

    static func load(repository: BockMediaRepository) async -> SearchBrowseFeed {
        async let historyTask = loadHistory(repository: repository)
        async let analyticsTask = analyticsWithTimeout(repository: repository)
        async let playlistsTask = loadPlaylists(repository: repository)
        async let smartTask = try? await repository.smartPlaylists()
        async let genresTask = try? await repository.genres(limit: 16)
        async let newReleasesTask = try? await repository.recentAlbums(limit: 12)
        async let dashboardTask = loadDashboard(repository: repository)

        let history = await historyTask?.items ?? []
        let analytics = await analyticsTask
        let allPlaylists = await playlistsTask?.items ?? []
        let smartPlaylists = await smartTask ?? []
        let genres = await genresTask ?? []
        let newReleases = await newReleasesTask?.items ?? []
        let dashboard = await dashboardTask

        let playlistByName = Dictionary(
            allPlaylists.map { ($0.name.lowercased(), $0) },
            uniquingKeysWith: { _, last in last }
        )
        let playlistById = Dictionary(
            allPlaylists.map { ($0.id, $0) },
            uniquingKeysWith: { _, last in last }
        )
        var usedArtPaths = Set<String>()

        func claimArt(_ path: String?) -> String? {
            guard let path, !path.isEmpty, !usedArtPaths.contains(path) else { return nil }
            usedArtPaths.insert(path)
            return path
        }

        func playlistCard(_ pl: PlaylistSummary, artPath: String?, subtitle: String) -> HomeCard {
            HomeCard(
                id: "pl-\(pl.id)",
                title: pl.name,
                subtitle: subtitle,
                artPath: artPath,
                playlistId: pl.id,
                playTarget: .playlist(id: pl.id, name: pl.name),
                kind: .discover
            )
        }

        var recentPlaylistNames: [String] = []
        var artByPlaylist: [String: String] = [:]
        for row in history {
            guard let name = row.playlist?.trimmingCharacters(in: .whitespaces), !name.isEmpty else { continue }
            if !recentPlaylistNames.contains(where: { $0.caseInsensitiveCompare(name) == .orderedSame }) {
                recentPlaylistNames.append(name)
            }
            if let path = row.filepath {
                artByPlaylist[name.lowercased()] = artByPlaylist[name.lowercased()] ?? path
            }
        }

        let topGenres = Array(analytics?.topGenres.prefix(6) ?? [])
        let topArtists = analytics?.topArtists ?? []

        let genreMixes: [HomeCard] = topGenres.enumerated().compactMap { index, row in
            let genre = row.name ?? row.label ?? ""
            guard !genre.isEmpty else { return nil }
            let smart = smartPlaylists.first(where: { sp in
                !HomeFeedRules.isDailyMixName(sp.name) && sp.name.localizedCaseInsensitiveContains(genre)
            })
            let artPath = claimArt(HomeFeedRules.artPathForGenreDistinct(history, genre: genre, used: usedArtPaths))
                ?? claimArt(HomeFeedRules.artPathForArtistDistinct(
                    history,
                    artist: HomeFeedRules.topArtistForGenre(history, genre: genre)
                        ?? topArtists[safe: index]?.name ?? topArtists[safe: index]?.label ?? "",
                    used: usedArtPaths
                ))
                ?? claimArt(HomeFeedRules.nextDistinctArtPath(history, used: usedArtPaths))

            if let linkedId = smart?.playlistId, let pl = playlistById[linkedId] {
                return playlistCard(pl, artPath: artPath, subtitle: "\(genre) mix")
            }
            if let named = allPlaylists.first(where: { HomeFeedRules.isGenreMixPlaylistName($0.name, genre: genre) }) {
                return playlistCard(named, artPath: artPath, subtitle: "\(genre) mix")
            }
            guard let seed = HomeFeedRules.topArtistForGenre(history, genre: genre)
                ?? topArtists[safe: index]?.name ?? topArtists[safe: index]?.label else { return nil }
            return HomeCard(
                id: "mix-\(genre)",
                title: "\(genre) Mix",
                subtitle: "Based on your listening",
                artPath: artPath,
                playlistId: nil,
                playTarget: .artist(name: seed),
                kind: .topMixes
            )
        }

        var recentAlbumCards: [HomeCard] = []
        var seenAlbums = Set<String>()
        for row in history {
            guard let album = row.album?.trimmingCharacters(in: .whitespaces), !album.isEmpty else { continue }
            let key = "\(album.lowercased())|\(row.artist?.lowercased() ?? "")"
            guard seenAlbums.insert(key).inserted else { continue }
            recentAlbumCards.append(HomeCard(
                id: "album-\(key)",
                title: album,
                subtitle: row.artist ?? "Recently played",
                artPath: claimArt(row.filepath),
                playlistId: nil,
                playTarget: .album(name: album, artist: row.artist),
                kind: .jumpBackIn
            ))
            if recentAlbumCards.count >= 4 { break }
        }

        let discoverCards = allPlaylists
            .filter { HomeFeedRules.isDiscoverName($0.name) }
            .sorted { parseSortDate($0.createDate) > parseSortDate($1.createDate) }
            .prefix(4)
            .map { pl in
                playlistCard(pl, artPath: claimArt(artByPlaylist[pl.name.lowercased()]), subtitle: "Discover")
            }

        let recentPlaylistCards = recentPlaylistNames.prefix(4).compactMap { name -> HomeCard? in
            guard let pl = playlistByName[name.lowercased()] else { return nil }
            return playlistCard(pl, artPath: claimArt(artByPlaylist[name.lowercased()]), subtitle: "Played recently")
        }

        let dashboardCards = dashboardRecentCards(
            dashboard,
            playlistByName: playlistByName,
            artByPlaylist: artByPlaylist,
            claimArt: claimArt
        )

        let pickedForYou = (dashboardCards + genreMixes + recentAlbumCards + discoverCards + recentPlaylistCards)
            .uniqued(by: \.id)
            .prefix(12)
            .map { $0 }

        return SearchBrowseFeed(pickedForYou: pickedForYou, newReleases: newReleases, genres: genres)
    }

    static func suggestOneChar(repository: BockMediaRepository, query: String) async -> [SearchSuggestion] {
        guard let artists = try? await repository.artists(page: 1, limit: 6, search: query) else { return [] }
        return artists.items.map { artist in
            SearchSuggestion(
                kind: .artist,
                title: artist.name,
                subtitle: "\(artist.tracks) songs",
                hitId: nil,
                path: nil,
                artist: nil,
                album: nil
            )
        }
    }

    static func suggestionsFromResponse(_ response: SearchResponse) -> [SearchSuggestion] {
        var out: [SearchSuggestion] = []
        for hit in response.artists.prefix(2) {
            guard let title = hit.displayName, !title.isEmpty else { continue }
            out.append(SearchSuggestion(kind: .artist, title: title, subtitle: "Artist", hitId: hit.id, path: hit.path, artist: hit.artist, album: nil))
        }
        for hit in response.albums.prefix(2) {
            guard let title = hit.displayName, !title.isEmpty else { continue }
            out.append(SearchSuggestion(kind: .album, title: title, subtitle: hit.artist, hitId: hit.id, path: hit.path, artist: hit.artist, album: hit.album))
        }
        for hit in response.playlists.prefix(2) {
            out.append(SearchSuggestion(kind: .playlist, title: hit.name ?? "", subtitle: "Playlist", hitId: hit.id, path: hit.path, artist: nil, album: nil))
        }
        for hit in response.songs.prefix(2) {
            out.append(SearchSuggestion(kind: .song, title: hit.title ?? hit.name ?? "", subtitle: hit.artist, hitId: hit.id, path: hit.path, artist: hit.artist, album: hit.album))
        }
        return Array(out.prefix(8))
    }

    static func playTarget(for suggestion: SearchSuggestion) -> PlayTarget? {
        switch suggestion.kind {
        case .artist: return .artist(name: suggestion.title)
        case .album: return .album(name: suggestion.title, artist: suggestion.artist)
        case .playlist:
            guard let id = suggestion.hitId else { return nil }
            return .playlist(id: id, name: suggestion.title)
        case .song:
            guard let path = suggestion.path else { return nil }
            return .song(path: path, title: suggestion.title)
        }
    }

    private static func loadHistory(repository: BockMediaRepository) async -> StreamHistoryResponse? {
        if let cached = SessionDataStore.peekHistory() { return cached }
        return try? await repository.streamHistory(limit: historyLimit)
    }

    private static func loadPlaylists(repository: BockMediaRepository) async -> PlaylistsResponse? {
        let memberKey = ActiveProfileStore.activeMemberId() ?? ""
        if let cached = SessionDataStore.peekPlaylists(memberKey: memberKey) { return cached }
        return try? await repository.playlists(limit: 200)
    }

    private static func loadDashboard(repository: BockMediaRepository) async -> DashboardQuickResponse? {
        if let cached = SessionDataStore.peekDashboard() { return cached }
        return try? await repository.dashboardQuick()
    }

    private static func analyticsWithTimeout(repository: BockMediaRepository) async -> AnalyticsResponse? {
        if let cached = SessionDataStore.peekAnalytics() { return cached }
        return await withTaskGroup(of: AnalyticsResponse?.self) { group in
            group.addTask { try? await repository.analytics() }
            group.addTask {
                try? await Task.sleep(nanoseconds: 4_000_000_000)
                return nil
            }
            let first = await group.next()
            group.cancelAll()
            return first ?? nil
        }
    }

    private static func dashboardRecentCards(
        _ dashboard: DashboardQuickResponse?,
        playlistByName: [String: PlaylistSummary],
        artByPlaylist: [String: String],
        claimArt: (String?) -> String?
    ) -> [HomeCard] {
        guard let dashboard else { return [] }
        return dashboard.recent.compactMap { item -> HomeCard? in
            if let playlistName = item.playlist?.trimmingCharacters(in: .whitespaces), !playlistName.isEmpty,
               let pl = playlistByName[playlistName.lowercased()] {
                return HomeCard(
                    id: "dash-pl-\(pl.id)",
                    title: pl.name,
                    subtitle: "Recently played",
                    artPath: claimArt(artByPlaylist[playlistName.lowercased()] ?? item.path),
                    playlistId: pl.id,
                    playTarget: .playlist(id: pl.id, name: pl.name),
                    kind: .discover
                )
            }
            guard let path = item.path, let title = item.track?.trimmingCharacters(in: .whitespaces), !title.isEmpty else { return nil }
            return HomeCard(
                id: "dash-\(path)",
                title: title,
                subtitle: item.artist ?? "Recently played",
                artPath: claimArt(path),
                playlistId: nil,
                playTarget: .song(path: path, title: title),
                kind: .discover
            )
        }
    }

    private static func parseSortDate(_ value: String?) -> Date {
        guard let value, !value.isEmpty else { return .distantPast }
        let iso = ISO8601DateFormatter()
        iso.formatOptions = [.withInternetDateTime, .withFractionalSeconds]
        if let d = iso.date(from: value) { return d }
        iso.formatOptions = [.withFullDate]
        if let d = iso.date(from: value) { return d }
        let f = DateFormatter()
        f.dateFormat = "yyyy-MM-dd"
        return f.date(from: value) ?? .distantPast
    }
}

private extension Array {
    subscript(safe index: Int) -> Element? { indices.contains(index) ? self[index] : nil }
}

private extension Sequence {
    func uniqued<Key: Hashable>(by key: (Element) -> Key) -> [Element] {
        var seen = Set<Key>()
        return filter { seen.insert(key($0)).inserted }
    }
}
